Source/ProjectRimFactory/SAL3/Things/Assemblers/Special/Building_Assembler_Learning.cs

```diff
@@ -1,105 +1,114 @@
-            }
-        }
-        public override string GetInspectString()
-        {
-            StringBuilder stringBuilder = new StringBuilder();
-            stringBuilder.AppendLine(base.GetInspectString());
-            if (currentBillReport != null)
-            {
-                stringBuilder.AppendLine("SALCurrentProductionSpeed".Translate(currentBillReport.bill.recipe.label,ProductionSpeedFactor.ToStringPercent()));
-            }
-            return stringBuilder.ToString().TrimEndNewlines();
-        }
-
-        protected override void PostProcessRecipeProduct(Thing thing)
-        {
-        }
-
-        protected override IEnumerable<FloatMenuOption> GetDebugOptions()
-        {
-            foreach (FloatMenuOption option in base.GetDebugOptions())
-            {
-                yield return option;
-            }
-            yield return new FloatMenuOption("View active skills", () => {
-                manager.TrimUnnecessaryFactors();
-                StringBuilder stringBuilder = new StringBuilder();
-                stringBuilder.AppendLine("Active skills ranked descending:");
-                RecipeDef[] keys = manager.factors.Keys.ToArray();
-                WorkSpeedFactorEntry[] values = manager.factors.Values.ToArray();
-                Array.Sort(values, keys);
-                for (int i = keys.Length - 1; i >= 0; i--)
-                {
-                    stringBuilder.AppendLine($"{keys[i].LabelCap}: {values[i].FactorFinal + FactorOffset}");
-                }
-                Log.Message(stringBuilder.ToString());
-            });
-        }
-        public override void ExposeData()
-        {
-            Scribe_Deep.Look(ref manager, "manager");
-            base.ExposeData();
-        }
-    }
-}
+﻿using RimWorld;
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using UnityEngine;
+using Verse;
+using HarmonyLib;
+using ProjectRimFactory.Common;
+using ProjectRimFactory.Common.HarmonyPatches;
+
+namespace ProjectRimFactory.SAL3.Things.Assemblers.Special
+{
+    public class Building_Assembler_Learning : Building_SmartAssembler, ISetQualityDirectly
+    {
+        public float FactorOffset => modExtension_LearningAssembler?.MinSpeed ?? 0.5f;
+
+        public float MaxSpeed
+        {
+            get
+            {
+                var effectiveMaxSpeed = modExtension_LearningAssembler?.MaxSpeed ?? float.PositiveInfinity;
+
+                return effectiveMaxSpeed - FactorOffset;
+            }
+        }
+
+        public float Progress => Mathf.Clamp01(manager.GetFactorFor(currentBillReport.bill.recipe) / MaxSpeed);
+
+        /// <summary>
+        /// Maps the 0..1 progress to -.1..1.1 scale
+        /// Slightly favors low quality at low progress, and high quality at high progress.
+        /// </summary>
+        public float NormalizedProgress => (Progress * 1.2f) - 0.1f;
```

Comment:
I'm not entirely certain about this part. This is meant to make the curve more pronounced at the edges, tendency towards poorer quality at 0 progress, and tendency towards better quality at 1 progress.